File: Ai/eightPuzzle.py
import heapq

# Goal state
GOAL = ((1, 2, 3),
        (4, 5, 6),
        (7, 8, 0))   # 0 represents blank (_)


# Manhattan Distance Heuristic
def manhattan(state):
    distance = 0
    for i in range(3):
        for j in range(3):
            value = state[i][j]
            if value != 0:
                goal_x = (value - 1) // 3
                goal_y = (value - 1) % 3
                distance += abs(i - goal_x) + abs(j - goal_y)
    return distance


# Find position of blank (0)
def find_blank(state):
    for i in range(3):
        for j in range(3):
            if state[i][j] == 0:
                return i, j


# Generate neighbors
def get_neighbors(state):
    neighbors = []
    x, y = find_blank(state)

    moves = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # up, down, left, right

    for dx, dy in moves:
        nx, ny = x + dx, y + dy
        if 0 <= nx < 3 and 0 <= ny < 3:
            new_state = [list(row) for row in state]
            new_state[x][y], new_state[nx][ny] = new_state[nx][ny], new_state[x][y]
            neighbors.append(tuple(tuple(row) for row in new_state))

    return neighbors


# A* Algorithm
def a_star(start):
    open_list = []
    heapq.heappush(open_list, (0, start))

    came_from = {}
    g_cost = {start: 0}

    while open_list:
        _, current = heapq.heappop(open_list)

        if current == GOAL:
            return reconstruct_path(came_from, current)

        for neighbor in get_neighbors(current):
            temp_g = g_cost[current] + 1

            if neighbor not in g_cost or temp_g < g_cost[neighbor]:
                g_cost[neighbor] = temp_g
                f_cost = temp_g + manhattan(neighbor)
                heapq.heappush(open_list, (f_cost, neighbor))
                came_from[neighbor] = current

    return None


# Reconstruct path
def reconstruct_path(came_from, current):
    path = [current]
    while current in came_from:
        current = came_from[current]
        path.append(current)
    path.reverse()
    return path


# Print state
def print_state(state):
    for row in state:
        print(" ".join(str(x) if x != 0 else "_" for x in row))
    print()


# Initial state
start_state = ((1, 0, 2),
               (4, 5, 3),
               (7, 8, 6))


# Run A*
solution = a_star(start_state)

if solution:
    print("Solution found in", len(solution) - 1, "moves\n")
    for step in solution:
        print_state(step)
else:
    print("No solution found")

File: Ai/eightPuzzle.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <vector>
#include <cmath>
#include <set>

using namespace std;

typedef tuple<int, int, int, int, int, int, int, int, int> State;

const State GOAL = make_tuple(1, 2, 3, 4, 5, 6, 7, 8, 0);

// Convert state to string for hashing
string stateToString(const State& state) {
    string result;
    for (int i = 0; i < 9; i++) {
        result += to_string(get<i>(state)) + ",";
    }
    return result;
}

// Manhattan Distance Heuristic
int manhattan(const State& state) {
    int distance = 0;
    for (int i = 0; i < 9; i++) {
        int value = get<i>(state);
        if (value != 0) {
            int current_x = i / 3;
            int current_y = i % 3;
            int goal_x = (value - 1) / 3;
            int goal_y = (value - 1) % 3;
            distance += abs(current_x - goal_x) + abs(current_y - goal_y);
        }
    }
    return distance;
}

// Find position of blank (0)
int findBlank(const State& state) {
    for (int i = 0; i < 9; i++) {
        if (get<i>(state) == 0) {
            return i;
        }
    }
    return -1;
}

// Generate neighbors
vector<State> getNeighbors(const State& state) {
    vector<State> neighbors;
    int blank_pos = findBlank(state);
    int x = blank_pos / 3;
    int y = blank_pos % 3;

    int moves[][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    for (auto& move : moves) {
        int nx = x + move[0];
        int ny = y + move[1];
        if (nx >= 0 && nx < 3 && ny >= 0 && ny < 3) {
            State new_state = state;
            int new_pos = nx * 3 + ny;
            swap(get<blank_pos>(new_state), get<new_pos>(new_state));
            neighbors.push_back(new_state);
        }
    }
    return neighbors;
}

// Reconstruct path
vector<State> reconstructPath(unordered_map<string, string>& came_from, const State& current) {
    vector<State> path;
    string current_str = stateToString(current);
    path.push_back(current);

    while (came_from.find(current_str) != came_from.end()) {
        // Convert string back to state (simplified - just for tracking)
        current_str = came_from[current_str];
    }
    return path;
}

// Print state
void printState(const State& state) {
    for (int i = 0; i < 9; i++) {
        if (i != 0 && i % 3 == 0) cout << "\n";
        cout << (get<i>(state) == 0 ? "_" : to_string(get<i>(state))) << " ";
    }
    cout << "\n\n";
}

// A* Algorithm
vector<State> aStar(const State& start) {
    priority_queue<pair<int, State>, vector<pair<int, State>>, greater<pair<int, State>>> open_list;
    open_list.push({manhattan(start), start});

    unordered_map<string, string> came_from;
    unordered_map<string, int> g_cost;
    g_cost[stateToString(start)] = 0;

    while (!open_list.empty()) {
        auto [f, current] = open_list.top();
        open_list.pop();

        if (current == GOAL) {
            vector<State> result;
            result.push_back(current);
            return result;
        }

        vector<State> neighbors = getNeighbors(current);
        int current_g = g_cost[stateToString(current)];

        for (const State& neighbor : neighbors) {
            int temp_g = current_g + 1;
            string neighbor_str = stateToString(neighbor);

            if (g_cost.find(neighbor_str) == g_cost.end() || temp_g < g_cost[neighbor_str]) {
                g_cost[neighbor_str] = temp_g;
                int f_cost = temp_g + manhattan(neighbor);
                open_list.push({f_cost, neighbor});
                came_from[neighbor_str] = stateToString(current);
            }
        }
    }

    return vector<State>();
}

int main() {
    State start_state = make_tuple(1, 0, 2, 4, 5, 3, 7, 8, 6);

    vector<State> solution = aStar(start_state);

    if (!solution.empty()) {
        cout << "Solution found in " << solution.size() - 1 << " moves\n\n";
        for (const State& step : solution) {
            printState(step);
        }
    } else {
        cout << "No solution found\n";
    }

    return 0;
}
